experimental/Schemes/Tjurina.jl

```diff
@@ -0,0 +1,770 @@
+    try                                                                                         #TODO: simplify after vec_dim is improved
+      return milnor_number(X) != -1
+    catch
+      return false
+    end
+  end
+  if equivalence == :contact
+    return tjurina_number(f) != -1
+  end
+end
+
+
+
+
+@doc raw"""
+    is_finitelydetermined(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+
+Return if the hypersurface germ 'X' is finitely determined w.r.t. ':right' or ':contact' equivalence. 
+By default computes w.r.t. contact equivalence.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> f = x^2 - y^2;
+
+julia> X = HypersurfaceGerm(AffineScheme(quo(R, ideal(R, f))[1]), [0, 0]);
+
+julia> is_finitelydetermined(X)
+true
+
+julia> is_finitelydetermined(X, :right)
+true
+```
+"""
+function is_finitelydetermined(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+  f = defining_ideal(X)[1]
+  return is_finitelydetermined(f, equivalence)
+end
+
+
+
+###############################################################################
+@doc raw"""
+    determinacy_bound(f::MPolyLocRingElem)
+    determinacy_bound(f::MPolyLocRingElem, equivalence::Symbol = :contact)
+
+Compute some determinacy bound of 'f' w.r.t. ':right' or ':contact' equivalence.
+Return -1 if not finitely determined. 
+By default computes w.r.t. contact equivalence.
+This computation is based on the Milnor number respectively Tjurina number.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> L,_  = localization(R, complement_of_point_ideal(R, [0, 0]));
+
+julia> f = L(x^3 - y^2);
+
+julia> determinacy_bound(f)
+3
+
+julia> determinacy_bound(f, :right)
+3
+```
+"""
+function determinacy_bound(f::MPolyLocRingElem, equivalence::Symbol = :contact)
+  equivalence == :right || equivalence == :contact || error("Equivalence typ must be ':right' or ':contact'.")
+  ord_f = order_as_series(f)
+  ## if the order of f is 1, then f is right and contact equivalent to its 1-jet (smooth case)
+  ord_f != 1 || return 1
+  is_finitelydetermined(f, equivalence) || return -1
+  if equivalence == :right
+    if ord_f == 0     
+      ## A unit has the same right-determinacy as the power series without the constant term.
+      ## Therefore remove constant term.
+      R = base_ring(parent(f))
+      shift, _ = base_ring_shifts(parent(f))
+      a = shift(numerator(f))
+      b = shift(denominator(f)) 
+      L, _ = localization(R, complement_of_point_ideal(R, [coefficient_ring(R)(0) for i = 1:ngens(R)]))
+      f_shifted = L(a//b) 
+      f_new = f_shifted-(constant_coefficient(a)//constant_coefficient(b))
+      return determinacy_bound(f_new, equivalence)
+    end
+    X = HypersurfaceGerm(quo(parent(f), ideal(parent(f), f))[1])
+    k = milnor_number(X)
+  else  ## equivalence == :contact
+    ord_f != 0 || return 0
+    k = tjurina_number(f)
+  end
+  characteristic(parent(f)) != 0 || return k + 1
+  characteristic(parent(f)) == 0 || return 2*k - ord_f + 2
```

Comment:
This seems unnecessarily clever to me. I find this much easier to understand (hopefully I got the order right ... if not I think it kinda underlines my point ;-) )
```suggestion
  return characteristic(parent(f)) == 0 ? k + 1 : 2*k - ord_f + 2
```